### PR TITLE
added EIP712

### PR DESCRIPTION
**Pull request by BChainbuddy**
# CHANGES

I added a new EIP712 contract implemented on transfer function. The contract is pretty basic. The owner can mint the tokens. The users can then call the transfer function to transfer parts of their balances. The implementation includes everything necessary to be EIP712.

### DOMAIN SEPARATOR
```solidity
    bytes32 private DOMAIN_SEPARATOR;
    constructor(address _addr) Ownable(_addr) {
            require(_addr.code.length > 0, "Not a contract");
            DOMAIN_SEPARATOR = keccak256(
                abi.encode(
                    keccak256(
                        "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
                    ),
                    keccak256(bytes("eip712transfer")),
                    keccak256("\x01"), // Ethereum prefix for message format
                    block.chainid,
                    address(this)
                )
            );
        }
```

### TRANSACTION HASH
```solidity
function hashTransferRequest(
        TransferRequest memory _transfer
    ) internal view returns (bytes32) {
        return
            keccak256(
                abi.encodePacked(
                    "\x19\x01", // Ethereum prefix for message format
                    DOMAIN_SEPARATOR,
                    keccak256(
                        abi.encode(
                            _transfer.from,
                            _transfer.to,
                            _transfer.amount
                        )
                    )
                )
            );
    }
```

### SIGNATURE VERIFICATION
```solidity
    function verifySignature(
        address signer,
        bytes32 transferRequestHash,
        bytes calldata signature
    ) internal pure returns (bool) {
        return signer == transferRequestHash.recover(signature);
    }
```

### AND LASTLY THE TRANSFER FUNCTION WHICH FOLLOWS THE EIP712
```solidity
function transfer(
        address from,
        address to,
        uint256 amount,
        bytes calldata signature
    ) external {
        require(from == msg.sender, "Address not allowed to transfer");
        TransferRequest memory transferrequest = TransferRequest(
            from,
            to,
            amount
        );
        bytes32 dataHash = hashTransferRequest(transferrequest);
        require(
            verifySignature(from, dataHash, signature),
            "Invalid signature"
        );
        balance[from] -= amount;
        balance[to] += amount;
    }
```

### CODE REVIEW

**Comment by wiasliaw** [comment]
Hi, @BChainbuddy

Great work. I would like to ask for some change.

1. This is the repo to demo how to implement EIP-1271. Therefore, you should implement a contract called `CounterEIP712.sol` and set a smart contract wallet as owner. There's an simple smart contract wallet `Wallet.sol` here.

2. Test the implementation.

You can follow these contracts for details:
1. `Counter.sol`: signature signed by wallet directly.
2. `CounterEIP191.sol`: signature following EIP191 standard.
3. `Counter.t.sol`: testing `Counter.sol`.
4. `CounterEIP191.t.sol`: testing `CounterEIP191.sol`.

**Comment by BChainbuddy** [comment]
Hello,

I closely followed you guidelines and made modifications to the EIP712. Please let me know if I need to correct anything.

**I modified the whole contract to match EIP191 as a number counter. And also renamed the file.**
```solidity
    uint256 public number;
    function setNumber(uint256 _newNumber, bytes calldata signature) external {
        bytes32 dataHash = hashNewNumber(_newNumber);
        require(
            IERC1271(owner()).isValidSignature(dataHash, signature) ==
                IERC1271.isValidSignature.selector
        );
        number = _newNumber;
    }

```

**I implemented EIP-1271**
```solidity
    import {IERC1271} from "./interface/IERC1271.sol";
    function setNumber(uint256 _newNumber, bytes calldata signature) external {
        bytes32 dataHash = hashNewNumber(_newNumber);
        require(
            IERC1271(owner()).isValidSignature(dataHash, signature) ==
                IERC1271.isValidSignature.selector
        );
        number = _newNumber;
    }
```
**I also created the test contract for CounterEIP712**
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import "forge-std/Test.sol";
import "openzeppelin-contracts/contracts/utils/cryptography/MessageHashUtils.sol";

import {CounterEIP712} from "../src/CounterEIP712.sol";
import {Wallet} from "../src/Wallet.sol";

contract TestCounterEIP712 is Test {
    address private eoa;
    uint256 private privkey;

    Wallet private w;
    CounterEIP712 private c;

    function setUp() external {
        (eoa, privkey) = makeAddrAndKey("EOA");
        vm.deal(eoa, 100 ether);

        vm.startPrank(eoa);
        w = new Wallet();
        c = new CounterEIP712(address(w));
        vm.stopPrank();
    }

    function testSetNumber(uint256 expectResult) external {
        bytes32 hash = c.hashNewNumber(expectResult);
        (uint8 v, bytes32 r, bytes32 s) = vm.sign(privkey, hash);

        vm.startPrank(eoa);
        c.setNumber(expectResult, bytes.concat(r, s, bytes1(v)));
        vm.stopPrank();
        assertEq(c.number(), expectResult);
    }
}

```

File: src/CounterEIP712.sol [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import {IERC1271} from "./interface/IERC1271.sol";
+
+/**
+ * @title CounterEIP712
+ * @dev A smart contract that allows secure updates of a number using the EIP-712 standard.
+ * The contract is owned by an address that has the authority to verify and set the number.
+ */
+contract CounterEIP712 is Ownable {
+    // EIP-712 domain separator
+    bytes32 private DOMAIN_SEPARATOR;
```
Comment (wiasliaw) [inline src/CounterEIP712.sol]: `bytes32 private immutable DOMAIN_SEPARATOR`

File: src/CounterEIP712.sol [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * @dev Constructor to initialize the contract owner and set the EIP-712 domain separator.
+     * @param _addr The address of a contract to be used as a domain separator.
+     */
+    constructor(address _addr) Ownable(_addr) {
+        require(_addr.code.length > 0, "Not a contract");
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                keccak256(
```
Comment (wiasliaw) [inline src/CounterEIP712.sol]: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/94697be8a3f0dfcd95dfb13ffbd39b5973f5c65d/contracts/utils/cryptography/EIP712.sol#L37

make it a constant

**Review by wiasliaw — COMMENTED** [review]
There should be a function type hash for `setNumber`.

1. type hash

```solidity
bytes32 SETNUMBER_TYPEHASH = keccak256("SetNumber(uint256 value)");
```

2. struct hash

```solidity
bytes32 setNumber_structHash = keccak256(
  abi.encode(
    SETNUMBER_TYPEHASH,
    value
  )
);
```

3. get the digest

```solidity
bytes32 digest = keccak256(
  abi.encodePacked(
    "\x19\x01",
    DOMAIN_SEPARATOR,
    setNumber_structHash
  )
)
```

**Comment by BChainbuddy** [comment]
Hello, thanks for the heads up. I added the additional steps.

**Comment by wiasliaw** [comment]
LGTM!

**Pull request merged**